### PR TITLE
[ABC-003] : Added ProductService, tests for service, automapper

### PR DESCRIPTION
**Pull request by alexzaslavsky**

### CODE REVIEW

File: CrmWebApi.Tests/Services/ProductServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        public void Service_Throws_ArgumentNullException_When_Mapper_Is_Null()
+        {
+            Assert.That(() =>
+            {
+                sut = new ProductService(productRepository, null);
+            }, Throws.InstanceOf<ArgumentNullException>());
+        }
+
+        private IEnumerable<Product> GetAllProducts()
+        {
```
Comment (demchuknick6) [inline CrmWebApi.Tests/Services/ProductServiceTests.cs:79]: can be static

File: CrmWebApi/Repositories/ProductRepository.cs [hunk 2]
```diff
@@ -12,7 +14,15 @@ public class ProductRepository : IProductRepository
...
-            _databaseContext = databaseContext;
+            try
+            {
+                _databaseContext = databaseContext ?? throw new ArgumentNullException(nameof(databaseContext));
+            }
+            catch (Exception ex)
+            {
+                Trace.WriteLine(ex.Message);
+                throw;
+            }
```
Comment (demchuknick6) [inline CrmWebApi/Repositories/ProductRepository.cs:25]: you don't need to try-catch it here

File: CrmWebApi/Services/ProductService.cs [hunk 3]
```diff
@@ -13,8 +14,16 @@ public class ProductService : IProductService
...
+            try
+            {
+                _productRepository = productRepository ?? throw new ArgumentNullException(nameof(productRepository));
+                _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+            }
+            catch (Exception ex)
+            {
+                Trace.WriteLine(ex.Message);
+                throw;
+            }
```
Comment (demchuknick6) [inline CrmWebApi/Services/ProductService.cs:26]: same here

Comment (demchuknick6) [reply CrmWebApi/Services/ProductService.cs:26]: try-catch should be added for GetAll() and GetTheMostFrequentCategoryName() mathods, not for constructor

**Review by demchuknick6 — APPROVED** [review]

File: CrmWebApi.Tests/CrmWebApi.Tests.csproj [hunk 4]
```diff
@@ -9,6 +9,7 @@
   <ItemGroup>
     <PackageReference Include="EntityFrameworkCore.Testing.NSubstitute" Version="3.0.3" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.4" />
+    <PackageReference Include="Moq" Version="4.16.1" />
```
Comment (demchuknick6) [inline CrmWebApi.Tests/CrmWebApi.Tests.csproj:12]: remove moq

File: CrmWebApi.Tests/Services/ProductServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        IProductService sut;
+        IProductRepository productRepository;
+        IMapper mapper;
+
+        [SetUp]
+        public void Setup()
+        {
+            productRepository = Substitute.For<IProductRepository>();
+            productRepository.GetAll().Returns(GetAllProducts());
```
Comment (demchuknick6) [inline CrmWebApi.Tests/Services/ProductServiceTests.cs:25]: Returns(Enumerable.Empty<Product>)

File: CrmWebApi.Tests/Services/ProductServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        private static IEnumerable<Product> GetAllProducts()
+        {
+            var smartphone = new ProductCategory() { Id = 1, Name = "Smartphone" };
+            var laptop = new ProductCategory() { Id = 2, Name = "Laptop" };
+            var tv = new ProductCategory() { Id = 3, Name = "TV" };
+
+            return new Product[]
+                {
```
Comment (demchuknick6) [inline CrmWebApi.Tests/Services/ProductServiceTests.cs:86]: this can be removed then

File: CrmWebApi.Tests/Services/ProductServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+using CrmWebApi.ViewModels;
+using NSubstitute;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace CrmWebApi.Tests.Services
+{
+    internal class ProductServiceTests
```
Comment (demchuknick6) [inline CrmWebApi.Tests/Services/ProductServiceTests.cs:15]: test classes should be public. also, please add [TestFixture] attribute

File: CrmWebApi/Repositories/ProductRepository.cs [hunk 8]
```diff
@@ -12,7 +15,8 @@ public class ProductRepository : IProductRepository
 
         public ProductRepository(DatabaseContext databaseContext)
         {
-            _databaseContext = databaseContext;
+            GuardClauses.IsNotNull(databaseContext, nameof(databaseContext));
+            _databaseContext = databaseContext ?? throw new ArgumentNullException(nameof(databaseContext));
```
Comment (demchuknick6) [inline CrmWebApi/Repositories/ProductRepository.cs:19]: if your're throwing exception in Guard clause, then no need to do that on the next line:
```suggestion
            _databaseContext = databaseContext;
```

File: CrmWebApi/Services/ProductService.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            _mapper = mapper;
+        }
+
+        public IEnumerable<ProductViewModel> GetAll()
+        {
+            try
+            {
+                var products = _productRepository.GetAll();
+                return _mapper.Map<IEnumerable<ProductViewModel>>(products);
+
```
Comment (demchuknick6) [inline CrmWebApi/Services/ProductService.cs:31]: remove empty line

File: CrmWebApi.Helpers/GuardClauses.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+
+namespace CrmWebApi.Helpers
+{
+    public static class GuardClauses
```
Comment (demchuknick6) [inline CrmWebApi.Helpers/GuardClauses.cs:5]: Guard clauses could be done in scope of the next task

File: CrmWebApi/Repositories/ProductRepository.cs (hunk 8 above)
Comment (alexzaslavsky) [reply CrmWebApi/Repositories/ProductRepository.cs:19]: Yeah, I've seen that later, but haven't commited yet)
Bro, it's Sunday) Stop working :)

**Pull request merged**